### PR TITLE
Update actions/checkout

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
Many actions are outdated which triggers a lot of warnings. See [here](https://github.com/PennyLaneAI/pennylane-lightning/actions/runs/8570073242) for example.

**Description of the Change:**
Update some of the actions, those that do not introduce options requiring more in depth changes.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: .github/workflows/docker_linux_x86_64.yml [hunk 1]
```diff
@@ -43,7 +43,7 @@ jobs:
     steps:
 
       - name: Checkout
-        uses: actions/checkout@v3
+        uses: actions/checkout@v4
```
Comment (maliasadi) [inline .github/workflows/docker_linux_x86_64.yml:46]: @vincentmr  what would be the advantage of this change? is this required by any other task?

Comment (vincentmr) [reply .github/workflows/docker_linux_x86_64.yml:46]: Just some maintenance to remove a bunch of warnings.

**Review by AmintorDusko — APPROVED** [review]
Happy to see our actions up-to-date. Thank you, @vincentmr!

File: .github/workflows/docker_linux_x86_64.yml (hunk 1 above)
Comment (AmintorDusko) [reply .github/workflows/docker_linux_x86_64.yml:46]: Node.js 16 actions are deprecated, and several actions are warning about it.

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -40,6 +40,9 @@
 
 ### Improvements
 
+* Update GitHub actions versions.
```
Comment (maliasadi) [inline .github/CHANGELOG.md:43]:
```suggestion
* Upgrade GitHub actions versions from v3 to v4.
```

**Review by maliasadi — APPROVED** [review]
:raised_hands:

**Pull request merged**